spmd/compiler/fusion.py

```diff
@@ -0,0 +1,182 @@
+import logging
+from copy import deepcopy
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Iterable, List, Optional
+
+import torch
+import torch.fx as fx
+
+from .graph_utils import (
+    OP,
+    get_node_tensor_numel,
+    get_output_node,
+    graph_cleanup,
+    pretty_print_graph,
+)
+from .log_utils import rank0_debug
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+# enum for the supported fusion comm types
+class CommType(str, Enum):
+    allreduce = "allreduce_"
+    allgather = "allgather_"
+    broadcast = "broadcast_"
+    reducescatter = "reduce_scatter_"
+    scatter = "scatter_"
+
+
+@dataclass
+class FusionElement:
+    """This class tracks the nodes for a DTensor expanded communication collective in the graph"""
+
+    in_graph: bool = False
+    comm_type: Optional[CommType] = None
+    node_list: Optional[List[fx.Node]] = field(default_factory=lambda: [])  # type: ignore
+    size: int = 0
+    prev_node: Optional[fx.Node] = None  # node that was before start of section
+    next_node: Optional[fx.Node] = None  # node that was after end of section
+    processed: bool = False
+    output_name: str = ""
+    clone_node: Optional[fx.Node] = None
+    comm_node: Optional[fx.Node] = None
+    wait_node: Optional[fx.Node] = None
+
+
+@dataclass
+class GraphInfo:
+    len: int = 0
+    global_buffer: Optional[fx.Node] = None
+    global_buffer_size: int = 0
+    output: Optional[fx.Node] = None
+    first: Optional[fx.Node] = None
+
+    def update_info(self, gm: fx.GraphModule) -> None:
+        """get the len, input and output nodes"""
+        graph_len = gm.graph._len
+        if not graph_len:
+            raise ValueError("Empty graph passed in....")
+        self.len = graph_len
+
+        nodelist = gm.graph.nodes
+
+        for i, node in enumerate(nodelist):
+            if node.op == OP.PLACEHOLDER:
+                self.first = node
+                break
+
+        self.output = get_output_node(gm)
+        assert (
+            self.output is not None
+        ), f"unable to locate output node in gm {gm.graph}"
+
+        rank0_debug(
+            logger,
+            f"updated graph_info - len = {self.len} input = {self.first}, output = {self.output}",
+        )
+
+
+def _insert_fusion_buffer_node(
+    gm: fx.GraphModule, insert_before_node: fx.Node, buffer_size: Iterable[int]
+) -> fx.Node:
+    """insert a torch.empty node in front of insert_before_node"""
+    with gm.graph.inserting_before(insert_before_node):
+        new_buffer_node = gm.graph.create_node(
+            OP.CALL_FUNCTION,
+            target=torch.empty,
+            args=tuple(buffer_size),  # TODO - need device
```

Comment:
Can you add the TODO comment on top of the LOC and have more details?

Reply:
yes - moved comment location to above, and expanded comment to note that we need to get the device from the underlying DTensor to ensure the buffer is placed on same device as DTensor.